Applications/FileIO/SWMM/SWMMInterface.cpp

```diff
@@ -0,0 +1,1248 @@
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ */
+
+#include "SwmmInterface.h"
+
+#include <utility>
+
+#include "BaseLib/FileTools.h"
+#include "BaseLib/StringTools.h"
+#include "GeoLib/GEOObjects.h"
+#include "GeoLib/Point.h"
+#include "GeoLib/Polyline.h"
+#include "GeoLib/Polygon.h"
+
+#include "MeshLib/Mesh.h"
+#include "MeshLib/Node.h"
+#include "MeshLib/Elements/Line.h"
+#include "MeshLib/Properties.h"
+
+#include "Applications/FileIO/CsvInterface.h"
+
+#include "Applications/ApplicationsLib/LogogSetup.h"
+
+const std::array<std::string,9> subcatchment_vars =
+{
+    "rainfall",
+    "snow depth",
+    "evaporation loss",
+    "infiltration losses",
+    "runoff rate",
+    "groundwater outflow",
+    "groundwater head",
+    "moisture content",
+    "concentration of pollutant"
+};
+
+const std::array<std::string,7> node_vars =
+{
+    "water depth",
+    "hydraulic head",
+    "volume of stored water",
+    "lateral inflow",
+    "total inflow",
+    "flow lost to flooding",
+    "concentration of pollutant"
+};
+
+const std::array<std::string,6> link_vars =
+{
+    "flow rate",
+    "flow depth",
+    "flow velocity",
+    "flow volume",
+    "fraction conduit/non-conduit",
+    "concentration of pollutant"
+};
+
+const std::array<std::string,15> system_vars =
+{
+    "air temperature",
+    "rainfall",
+    "snow depth",
+    "evaporation + infiltration loss",
+    "runoff flow",
+    "dry weather inflow",
+    "groundwater inflow",
+    "RDII inflow",
+    "direct inflow",
+    "total lateral inflow",
+    "flow lost to flooding",
+    "flow leaving through outfalls",
+    "volume of stored water",
+    "actual evaporation rate",
+    "potential evaporation rate"
+};
+
+std::array<std::size_t,4> const n_obj_params = { 8, 6, 5, 15 };
+
+SwmmInterface* SwmmInterface::create(std::string const& file_name)
+{
+    if (file_name.length() < 5)
+        return nullptr;
+
+    if (!(SwmmInterface::isSwmmInputFile(file_name) || SwmmInterface::isSwmmOutputFile(file_name)))
+        return nullptr;
+
+    std::string const base_name (file_name.substr(0, file_name.length() - 4));
+    SwmmInterface* swmm = new SwmmInterface(base_name);
+    if (swmm->convertSwmmInputToLineMesh())
+        return swmm;
+
+    ERR ("Error creating mesh from SWMM file.");
+    delete swmm;
+    return nullptr;
+}
+
+SwmmInterface::SwmmInterface(std::string const& swmm_base_name)
+: _base_name (swmm_base_name), _mesh(nullptr)
+{
+}
+
+SwmmInterface::~SwmmInterface()
+{
+    for (Subcatchment sc  : _subcatchments)
+        delete sc.outline;
+
+    for (GeoLib::Point* pnt : _subcatchment_points)
+        delete pnt;
+}
+
+bool SwmmInterface::isSwmmInputFile(std::string const& inp_file_name)
+{
+    std::string path (inp_file_name);
+    std::transform(path.begin(), path.end(), path.begin(), tolower);
```

Comment:
Yes, all of these are valid. Old Windows codes don't distinguish between capitalisation in file(names?) or at least extensions.